coderd/parameters.go

```diff
@@ -127,6 +147,129 @@ func (api *API) templateVersionDynamicParameters(rw http.ResponseWriter, r *http
+				Icon:         it.Icon,
+				Options:      make([]*previewtypes.ParameterOption, 0),
+				Validations:  make([]*previewtypes.ParameterValidation, 0),
+				Required:     it.Required,
+				Order:        int64(it.DisplayOrder),
+				Ephemeral:    it.Ephemeral,
+				Source:       nil,
+			},
+			// Always use the default, since we used to assume the empty string
+			Value:       previewtypes.StringLiteral(it.DefaultValue),
+			Diagnostics: nil,
+		}
+
+		if it.ValidationError != "" || it.ValidationRegex != "" || it.ValidationMonotonic != "" {
+			var reg *string
+			if it.ValidationRegex != "" {
+				reg = ptr.Ref(it.ValidationRegex)
+			}
+
+			var vMin *int64
+			if it.ValidationMin.Valid {
+				vMin = ptr.Ref(int64(it.ValidationMin.Int32))
+			}
+
+			var vMax *int64
+			if it.ValidationMax.Valid {
+				vMin = ptr.Ref(int64(it.ValidationMax.Int32))
+			}
+
+			var monotonic *string
+			if it.ValidationMonotonic != "" {
+				monotonic = ptr.Ref(it.ValidationMonotonic)
+			}
+
+			param.Validations = append(param.Validations, &previewtypes.ParameterValidation{
+				Error:     it.ValidationError,
+				Regex:     reg,
+				Min:       vMin,
+				Max:       vMax,
+				Monotonic: monotonic,
+			})
+		}
+
+		var protoOptions []*sdkproto.RichParameterOption
+		_ = json.Unmarshal(it.Options, &protoOptions) // Not going to make this fatal
+		for _, opt := range protoOptions {
+			param.Options = append(param.Options, &previewtypes.ParameterOption{
+				Name:        opt.Name,
+				Description: opt.Description,
+				Value:       previewtypes.StringLiteral(opt.Value),
+				Icon:        opt.Icon,
+			})
+		}
+
+		// Take the form type from the ValidateFormType function. This is a bit
+		// unfortunate we have to do this, but it will return the default form_type
+		// for a given set of conditions.
+		_, param.FormType, _ = provider.ValidateFormType(provider.OptionType(param.Type), len(param.Options), param.FormType)
+
+		param.Diagnostics = previewtypes.Diagnostics(param.Valid(param.Value))
+		params = append(params, param)
+	}
+
+	api.handleParameterWebsocket(rw, r, func(_ context.Context, values map[string]string) (*preview.Output, hcl.Diagnostics) {
+		for i := range params {
+			param := &params[i]
+			paramValue, ok := values[param.Name]
+			if ok {
+				param.Value = previewtypes.StringLiteral(paramValue)
+			} else {
+				param.Value = param.DefaultValue
+			}
+			param.Diagnostics = previewtypes.Diagnostics(param.Valid(param.Value))
+		}
+
+		return &preview.Output{
+				Parameters: params,
+			}, hcl.Diagnostics{
+				{
+					// Only a warning because the form does still work.
+					Severity: hcl.DiagWarning,
+					Summary:  "This template version is missing required metadata to support dynamic parameters.",
+					Detail:   "To restore full functionality, please re-import the terraform as a new template version.",
+				},
+			}
+	})
+}
+
+func (api *API) handleParameterWebsocket(rw http.ResponseWriter, r *http.Request, render previewFunction) {
```

Comment:
```suggestion
func (api *API) handleParameterWebsocket(rw http.ResponseWriter, r *http.Request, preview previewFunction) {
```

still think we should call it preview, especially since that's what you named the type

Reply:
It will collide with the package name :cry: